test/unit/math/rev/mat/fun/matrix_exp_action_test.cpp

```diff
@@ -0,0 +1,368 @@
+
+  Eigen::Matrix<double, N, M> res = handler.action(A, B);
+  MatrixType Ad(A);
+  MatrixType expa = stan::math::matrix_exp(Ad);
+  Eigen::Matrix<double, N, M> expb = expa * B;
+
+  for (int i = 0; i < N; ++i) {
+    for (int j = 0; j < M; ++j) {
+      EXPECT_FLOAT_EQ(res(i, j), expb(i, j));
+    }
+  }
+}
+
+TEST(MathMatrix, matrix_exp_action_matrix_transpose) {
+  using MatrixType = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>;
+  stan::math::matrix_exp_action_handler handler;
+  std::srand(1999);
+
+  constexpr int N = 10;
+  constexpr int M = 4;
+  Eigen::Matrix<double, N, N> A = Eigen::Matrix<double, N, N>::Random();
+  Eigen::Matrix<double, N, M> B = Eigen::Matrix<double, N, M>::Random();
+
+  Eigen::Matrix<double, N, M> res = handler.action(A.transpose(), B);
+  MatrixType Ad(A);
+  MatrixType expa = stan::math::matrix_exp(Ad).transpose();
+  Eigen::Matrix<double, N, M> expb = expa * B;
+
+  for (int i = 0; i < N; ++i) {
+    for (int j = 0; j < M; ++j) {
+      EXPECT_NEAR(res(i, j), expb(i, j), 1.e-6);
+    }
+  }
+}
+
+template <int N, int M>
+inline void test_matrix_exp_action_dv() {
+  using stan::math::value_of;
+  using stan::math::var;
+  stan::math::matrix_exp_action_handler handler;
+  std::srand(1999);
+
+  Eigen::Matrix<var, N, N> Av = Eigen::Matrix<var, N, N>::Random();
+  Eigen::Matrix<var, N, M> Bv = Eigen::Matrix<var, N, M>::Random();
+  std::vector<stan::math::var> Bvec(Bv.data(), Bv.data() + Bv.size());
+  std::vector<stan::math::var> Avec(Av.data(), Av.data() + Av.size());
+  Eigen::Matrix<double, N, N> A = value_of(Av);
+
+  // brute force
+  Eigen::Matrix<double, N, N> expA = stan::math::matrix_exp(A);
+  Eigen::Matrix<var, N, M> expAB;
+  for (int i = 0; i < N; ++i) {
+    for (int j = 0; j < M; ++j) {
+      expAB(i, j) = 0.0;
+      for (int k = 0; k < N; ++k) {
+        expAB(i, j) += expA(i, k) * Bv(k, j);
+      }
+    }
+  }
+
+  // matrix_exp_action
+  Eigen::Matrix<var, N, M> res_dv = stan::math::matrix_exp_action(A, Bv);
+  for (int l = 0; l < res_dv.size(); ++l) {
+    EXPECT_FLOAT_EQ(res_dv(l).val(), expAB(l).val());
+  }
+  // compare adjoints
+  std::vector<double> g, g0;
+  for (int l = 0; l < M; ++l) {
+    for (int k = 0; k < N; ++k) {
+      stan::math::set_zero_all_adjoints();
+      res_dv(k, l).grad(Bvec, g);
+      stan::math::set_zero_all_adjoints();
+      expAB(k, l).grad(Bvec, g0);
+      for (size_t j = 0; j < g.size(); ++j) {
+        EXPECT_FLOAT_EQ(g[j], g0[j]);
+      }
+    }
+  }
+
+  // test a single function of expA*B
+  var f = sum(res_dv);
+  var f0 = sum(expAB);
+  stan::math::set_zero_all_adjoints();
+  f.grad(Bvec, g);
+  stan::math::set_zero_all_adjoints();
+  f0.grad(Bvec, g0);
+  for (size_t j = 0; j < g.size(); ++j) {
+    EXPECT_FLOAT_EQ(g[j], g0[j]);
+  }
+}
+
+TEST(MathMatrix, matrix_exp_action_dv) {
+  test_matrix_exp_action_dv<1, 1>();
+  test_matrix_exp_action_dv<1, 5>();
+  test_matrix_exp_action_dv<5, 1>();
+  test_matrix_exp_action_dv<5, 5>();
+}
```

Comment:
If this doesn't slow down the code too much, consider using higher dimensions for matrices. Can you go up to 100 x 100?

Reply:
It's probably asking too much for a unit test. Such a test is more like in performance regime. I've added a 20x20 test though.